L1Trigger/L1CaloTrigger/plugins/L1TS2PFJetInputPatternWriter.cc

```diff
@@ -0,0 +1,305 @@
+  std::string outDir_;
+
+  // constants
+  unsigned nChan_;  // number of channels per quad
+  unsigned nQuad_;
+  unsigned nLink_;
+  unsigned nHeaderFrames_;
+  unsigned nPayloadFrames_;
+  unsigned nClearFrames_;
+  unsigned nFrame_;
+  unsigned nFrameFile_;
+  unsigned nEvents_;
+
+  // data arranged by link and frame
+  std::vector<std::vector<uint64_t>> data_;
+
+  // data valid flags (just one per frame for now)
+  std::vector<int> dataValid_;
+
+  // map of towers onto links/frames
+  std::map<int, int> map_;
+};
+
+//
+// constants, enums and typedefs
+//
+
+//
+// static data member definitions
+//
+
+//
+// constructors and destructor
+//
+L1TS2PFJetInputPatternWriter::L1TS2PFJetInputPatternWriter(const edm::ParameterSet& iConfig)
+    : pfToken_(consumes<std::vector<l1t::PFCandidate>>(iConfig.getParameter<edm::InputTag>("pfTag"))),
+      filename_(iConfig.getUntrackedParameter<std::string>("filename")),
+      outDir_(iConfig.getUntrackedParameter<std::string>("outDir")),
+      nChan_(iConfig.getUntrackedParameter<unsigned>("nChanPerQuad")),
+      nQuad_(iConfig.getUntrackedParameter<unsigned>("nQuads")),
+      nHeaderFrames_(iConfig.getUntrackedParameter<unsigned>("nHeaderFrames")),
+      nPayloadFrames_(iConfig.getUntrackedParameter<unsigned>("nPayloadFrames")),
+      nClearFrames_(iConfig.getUntrackedParameter<unsigned>("nClearFrames")) {
+  //now do what ever initialization is needed
+
+  // register what you consume and keep token for later access:
+
+  nFrame_ = 0;
+  nFrameFile_ = 0;
+  nEvents_ = 0;
+
+  nLink_ = nChan_ * nQuad_;
+  data_.resize(nLink_);
+  LogDebug("L1TDebug") << "Preparing for " << nLink_ << " links" << std::endl;
+}
+
+L1TS2PFJetInputPatternWriter::~L1TS2PFJetInputPatternWriter() {
+  // do anything here that needs to be done at desctruction time
+  // (e.g. close files, deallocate resources etc.)
+}
+
+//
+// member functions
+//
+
+// ------------ method called for each event  ------------
+void L1TS2PFJetInputPatternWriter::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  using namespace edm;
+
+  //count events
+  nEvents_++;
+
+  edm::Handle<std::vector<l1t::PFCandidate>> pfHandle;
+  iEvent.getByToken(pfToken_, pfHandle);
+  std::vector<l1t::PFCandidate> const& pfCands = *pfHandle;
+
+  std::vector<l1t::PFCandidate> pfPartsA;
+  std::vector<l1t::PFCandidate> pfPartsB;
+
+  for (auto const& pfc : pfCands) {
+    // select first two "small" regions for current fw
+    if (pfc.eta() >= 0 && pfc.eta() < eta_first_region_boundary && pfc.phi() >= 0 && pfc.phi() < phi_region_boundary)
+      pfPartsA.push_back(pfc);
+    if (pfc.eta() >= eta_first_region_boundary && pfc.eta() < eta_second_region_boundary && pfc.phi() >= 0 &&
+        pfc.phi() < phi_region_boundary)
+      pfPartsB.push_back(pfc);
+  }
+
+  if (pfPartsA.empty() && pfPartsB.empty())
+    return;
+
+  if (nFrame_ == 0 || nFrameFile_ == 0) {
+    //first empty frames
+    while (nFrameFile_ < n_first_empty_frames) {
```

Comment:
it seems that this triple loop can mostly be replaced by
dataValid_.resize(dataValid_.size() + n_first_empty_frames,1);
for (auto &v : data_) 
  v.resize(v.size() + n_first_empty_frames,0);
nFrame_ = n_first_empty_frames;
nFrameFile_ = n_first_empty_frames;

Reply:
Thanks @davidlange6 
Perhaps replacing of
```
   while (nFrameFile_ < n_first_empty_frames) {
      dataValid_.push_back(1);
```
with
```
dataValid_.resize(dataValid_.size() + n_first_empty_frames,1);
``` 
yes, but then replacing the filling of the `data_` vector at indices `iLink`, does not seem to be straight forward. 
So I would prefer to leave as is, and have the expert (@bundocka ) of this analyzer `L1TS2PFJetInputPatternWriter` come back to this address this in future PR.

Reply:
iLink just counts from 0 to nQuad_*nChan_ -1, no?